### PR TITLE
Make the install script support kylin v10

### PR DESCRIPTION
**Pull request by ErikJiang**
#### Proposed Changes ####

Add number type detection to VERSION_ID to avoid errors in Kylin V10 OS due to integer comparison.

#### Types of Changes ####
Bugfix

#### Linked Issues ####

https://github.com/k3s-io/k3s/issues/8206

#### User-Facing Change ####

```release-note
NONE
```

### CODE REVIEW

**Review by brandond — CHANGES_REQUESTED** [review]
As per the freedesktop spec, the field is allowed to contain non-numeric characters:
> https://www.freedesktop.org/software/systemd/man/os-release.html#VERSION_ID=
> A lower-case string (mostly numeric, no spaces or other characters outside of 0–9, a–z, ".", "_" and "-") identifying the operating system version, excluding any OS name information or release code name, and suitable for processing by scripts or usage in generated filenames.

Rather than hardcoding a fix for this specific distro, we should fix or rework any comparisons that expect it to be an integer.

**Comment by brandond** [comment]
If Kylin is intended to be compatible with EL8, they really should include an ID_LIKE field in their os-release to indicate that. I am not super comfortable pointing it at EL packages without some indication that these are intended to and will continue to work in the future.

**Comment by ErikJiang** [comment]
@brandond  🤝Thanks for the suggestion, I've updated it, please take a look at it.

**Comment by ErikJiang** [comment]
@brandond PTAL 🙏

**Comment by nodakai** [comment]
Two cents:

https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02

`${VERSION_ID-%%.*}` should evaluate to an empty string when `$VERSION_ID` isn't defined (my problem on a Debian node). Additionally, because the default value (after `-`) is an empty string, the expression acts the same when `$VERSION_ID` is initially set to an empty string (theoretical corner case).

The original code repeats the expression for Cent 7, 8 and Fedora but I'd assign the result to another temp variable like `sanitized_id` (DRY)

`isNumber` might only be required for the `-gt 36` comparison, as the `=` operator performs a string-to-string comparison and happily returns 1 (false) for inputs like `V10`

Also, it can probably be implemented with the builtin `case` syntax without needing to invoke `grep`

https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_04_05

```
% echo 'case "8" in [0-9]* | [0-9]*.[0-9]*) echo yes;; *) echo no; esac' | sh
yes
% echo 'case "36" in [0-9]* | [0-9]*.[0-9]*) echo yes;; *) echo no; esac' | sh
yes
% echo 'case "9.0" in [0-9]* | [0-9]*.[0-9]*) echo yes;; *) echo no; esac' | sh
yes
% echo 'case "" in [0-9]* | [0-9]*.[0-9]*) echo yes;; *) echo no; esac' | sh
no
% echo 'case "V10" in [0-9]* | [0-9]*.[0-9]*) echo yes;; *) echo no; esac' | sh
no
```

(I noticed the original code invokes `expr` for handling the SUSE version convention, I believe `case` can replace it as well)

Last but not least:

There's a check that runs **after** detecting a distro/version but is essentially distro/version agnostic. This is why my k3s installation succeeded on my Debian node despite an error in version detection.

https://github.com/k3s-io/k3s/blob/6aef26e94b57934185eb2e184ec75ce77dc5ca2b/install.sh#L548

If your installation doesn't have `/usr/share/selinux`, the shell function simply exits. Perhaps moving this check to the top of the function would benefit many users who don't use SELinux.

(I don't know if your distro has the SELinux directory or not)

File: install.sh [hunk 1]
```diff
@@ -444,6 +444,15 @@ download_hash() {
...
 
+# --- check if it is a number type  ---
+isNumber() {
+    if echo "${1}" | grep -Pq '^[0-9]+([.][0-9]+)?$'; then
+        return
+    else
+        return 1
+    fi
+}
+
```
Comment (dereknola) [inline install.sh:455]: Don't use grep, we can use sh builtin glob matching
```suggestion
# --- check if it is a number type  ---
is_number() {
  case $1 in
    ''|*[!0-9]*) return 1 ;;
    *) return 0 ;;
  esac
}

```

File: install.sh [hunk 2]
```diff
@@ -518,11 +527,11 @@ setup_selinux() {
         rpm_target=coreos
         rpm_site_infix=coreos
         package_installer=rpm-ostree
-    elif [ "${VERSION_ID%%.*}" = "7" ]; then
+    elif isNumber "${VERSION_ID}" && [ "${VERSION_ID%%.*}" = "7" ]; then
```
Comment (dereknola) [inline install.sh:530]: Standard format is underscores for sh functions. Also this addresses https://github.com/k3s-io/k3s/issues/8723 case.
```suggestion
    elif is_number "${VERSION_ID:-}" && [ "${VERSION_ID%%.*}" = "7" ]; then
```

File: install.sh [hunk 3]
```diff
@@ -518,11 +527,11 @@ setup_selinux() {
...
         rpm_target=el7
         rpm_site_infix=centos/7
         package_installer=yum
-    elif [ "${VERSION_ID%%.*}" = "8" ] || [ "${VERSION_ID%%.*}" -gt "36" ]; then
+    elif isNumber "${VERSION_ID}" && { [ "${VERSION_ID%%.*}" = "8" ] || [ "${VERSION_ID%%.*}" -gt "36" ]; }; then
```
Comment (dereknola) [inline install.sh:534]:
```suggestion
    elif is_number "${VERSION_ID:-}" && { [ "${VERSION_ID%%.*}" = "8" ] || [ "${VERSION_ID%%.*}" -gt "36" ]; }; then
```

**Review by dereknola — CHANGES_REQUESTED** [review]

**Comment by ErikJiang** [comment]
Thanks to @nodakai and @dereknola for their suggestions.

**Review by dereknola — APPROVED** [review]

File: install.sh (hunk 1 above)
Comment (nodakai) [reply install.sh]: I admit the shell glob I posted above was wrong but I suspect this version won't work as expected for `VERSION_ID` containing a decimal point which is quite common:

https://forums.centos.org/viewtopic.php?t=75942
https://gist.github.com/SwooshyCueb/4905ac15e3b5658f6c6907d9562af479

**Comment by brandond** [comment]
If you're still interested, would you mind rebasing this on top of current master?

**Comment by ErikJiang** [comment]
@brandond Thanks for the heads up. 🤝

**Review by brandond — CHANGES_REQUESTED** [review]
The other new requirement on master is updating the sha256sum of the install script. Thank you!

**Comment by ErikJiang** [comment]
I'm done.  @brandond

**Comment by ErikJiang** [comment]
@brandond PTAL 🙏

**Pull request closed without merging**